heka/plugins/s3splitfile/s3splitfile_input.go

```diff
@@ -0,0 +1,203 @@
+
+	input.stop = make(chan bool)
+	input.listChan = make(chan string, 1000)
+
+	return nil
+}
+
+func (input *S3SplitFileInput) Stop() {
+	close(input.stop)
+}
+
+func (input *S3SplitFileInput) Run(runner pipeline.InputRunner, helper pipeline.PluginHelper) error {
+	// Begin listing the files (either straight from S3 or from a cache)
+	// Write matching filenames on a "lister" channel
+	// Read from the lister channel:
+	//   - fetch the filename
+	//   - read records from it
+	//   - write them to a "reader" channel
+
+	var (
+		wg sync.WaitGroup
+		i  uint32
+	)
+
+	wg.Add(1)
+	go func() {
+		runner.LogMessage("Starting S3 list")
+		for r := range S3Iterator(input.bucket, input.S3BucketPrefix, input.schema) {
+			if r.Err != nil {
+				runner.LogError(fmt.Errorf("Error getting S3 list: %s", r.Err))
+			} else {
+				runner.LogMessage(fmt.Sprintf("Found: %s", r.Key.Key))
+				input.listChan <- r.Key.Key
+			}
+		}
+		// All done listing, close the channel
+		runner.LogMessage("All done listing. Closing channel")
+		close(input.listChan)
+		wg.Done()
+	}()
+
+	// Run a pool of concurrent readers.
+	for i = 0; i < input.S3WorkerCount; i++ {
+		wg.Add(1)
+		go input.fetcher(runner, &wg)
+	}
+	wg.Wait()
+
+	return nil
+}
+
+// TODO: handle "no such file"
+// TODO: use s3splitfile_common.ReadS3File()?
+func (input *S3SplitFileInput) readS3File(runner pipeline.InputRunner, s3Key string) (err error) {
+	runner.LogMessage(fmt.Sprintf("Preparing to read: %s", s3Key))
+
+	if input.bucket == nil {
+		runner.LogMessage(fmt.Sprintf("Dude, where's my bucket: %s", s3Key))
+		return
+	}
+
+	deliverer := runner.NewDeliverer(s3Key)
+	defer deliverer.Done()
+	sr := runner.NewSplitterRunner(s3Key)
+
+	rc, err := input.bucket.GetReader(s3Key)
+	if err != nil {
+		runner.LogError(fmt.Errorf("Error getting a reader: %s", err))
+	}
+	defer rc.Close()
+
+	runner.LogMessage(fmt.Sprintf("Reading messages from %s", s3Key))
+	for err == nil {
+		err = sr.SplitStream(rc, deliverer)
+	}
+
+	if err != io.EOF {
+		runner.LogError(fmt.Errorf("Error reading %s: %s", s3Key, err))
+	}
+
+	return
+}
+
+func (input *S3SplitFileInput) fetcher(runner pipeline.InputRunner, wg *sync.WaitGroup) {
+	var (
+		s3Key     string
+		startTime time.Time
+		duration  float64
+	)
+
+	ok := true
+	for ok {
+		select {
+		case s3Key, ok = <-input.listChan:
+			if !ok {
+				// Channel is closed => we're shutting down, exit cleanly.
+				// runner.LogMessage("Fetcher all done! shutting down.")
+				break
+			}
+
+			startTime = time.Now().UTC()
+			err := input.readS3File(runner, s3Key)
+			if err != nil && err != io.EOF {
+				runner.LogError(fmt.Errorf("Error reading %s: %s", s3Key, err))
+				continue
+			}
+			duration = time.Now().UTC().Sub(startTime).Seconds()
+			runner.LogMessage(fmt.Sprintf("Successfully fetched %s in %.2fs ", s3Key, duration))
+		}
+	}
+
+	wg.Done()
+}
+
+func init() {
+	pipeline.RegisterPlugin("S3SplitFileInput", func() interface{} {
+		return new(S3SplitFileInput)
+	})
+}
```

Comment:
It would be nice to have a ReportMsg interface to get some visibility into its operation.